src/ServantElm.hs

```diff
@@ -126,12 +157,44 @@ mkUrl segments =
           error
             "to implement - for captures, not needed now"
 
+elmTypeRefToDoc :: TypeRef -> Doc ann
+elmTypeRefToDoc = \case
+  RefPrim elmPrim -> elmPrimToDoc elmPrim
+  RefCustom (TypeName typeName) -> pretty typeName
+
+elmTypeParenDoc :: TypeRef -> Doc ann
+elmTypeParenDoc = parens . elmTypeRefToDoc
+
+elmPrimToDoc :: ElmPrim -> Doc ann
+elmPrimToDoc = \case
+  ElmUnit -> "()"
+  ElmNever -> "Never"
+  ElmBool -> "Bool"
+  ElmChar -> "Char"
+  ElmInt -> "Int"
+  ElmFloat -> "Float"
+  ElmString -> "String"
+  ElmTime -> "Posix"
+  ElmMaybe t -> "Maybe" <+> elmTypeParenDoc t
+  ElmResult l r -> "Result" <+> elmTypeParenDoc l <+> elmTypeParenDoc r
+  ElmPair a b -> lparen <> elmTypeRefToDoc a <> comma <+> elmTypeRefToDoc b <> rparen
+  ElmTriple a b c -> lparen <> elmTypeRefToDoc a <> comma <+> elmTypeRefToDoc b <> comma <+> elmTypeRefToDoc c <> rparen
+  ElmList l -> "List" <+> elmTypeParenDoc l
+
 mkTypeSignature :: Req ElmDefinition -> Doc ann
-mkTypeSignature _ =
-  (hsep . punctuate " ->") ("String" : catMaybes [toMsgType, returnType])
+mkTypeSignature request =
+  (hsep . punctuate " ->") ("String" : catMaybes [toMsgType, bodyType, returnType])
   where
+    elmTypeRef :: ElmDefinition -> Doc ann
+    elmTypeRef eDef = elmTypeRefToDoc $ definitionToRef eDef
     toMsgType :: Maybe (Doc ann)
-    toMsgType = Just . parens $ "Result Http.Error ()" <+> "-> msg"
+    toMsgType =
+      fmap mkMsgType $ request ^. reqReturnType
+      where
+        mkMsgType x = "(Result Http.Error" <+> parens (elmTypeRef x) <+> "-> msg)"
```

Comment:
Use `parens` instead of manually putting parentheses.

Reply:
done